cpmpy/expressions/globalconstraints.py

```diff
@@ -312,6 +312,146 @@ def value(self):
+        nb_nodes_by_level, num_mapping, transitions_by_level_normalized = self.normalize_layer_representation(nbl, tbl)
+        if len(transitions_by_level_normalized) > 2:
+            # decomposition with multiple transitions table and aux variables for the nodes
+            aux = [intvar(0, nb_nodes) for nb_nodes in nb_nodes_by_level[1:]]
+            for i in range(len(arr)):
+                transition_dummy = [[num_mapping[n], v, nb_nodes_by_level[i+1]] for n in nbl[i] for v in range(lb[i], ub[i] + 1) if
+                            (n, v) not in self.mapping]
+                if i != 0:
+                    transition_dummy += [[nb_nodes_by_level[i], v, nb_nodes_by_level[i+1]] for v in range(lb[i], ub[i] + 1)]
+                transitions_by_level_normalized[i] = transitions_by_level_normalized[i] + transition_dummy
+            tab_first = [x[1:] for x in transitions_by_level_normalized[0]]  # optimization for first level (one node)
+            return [aux[-1] == 0], [Table([arr[0], aux[0]], tab_first)] \
+                   + [Table([aux[i - 1], arr[i], aux[i]], transitions_by_level_normalized[i]) for i in
+                      range(1, len(arr))]
+            # defining constraints: aux and arr variables define a path in the augmented-with-negative-path-MDD
+            # constraining constraint: end of the path in accepting node
+        elif len(transitions_by_level_normalized) == 2:
+            # decomposition by unfolding into a table
+            tab = [[t_a[1], t_b[1]] for t_a in transitions_by_level_normalized[0] for t_b in
+                   transitions_by_level_normalized[1] if t_a[2] == t_b[0]]
+            return [Table(arr, tab)], []
+
+        elif len(transitions_by_level_normalized) == 1:
+            # decomposition to inDomain
+            return [InDomain(arr[0], [t[1] for t in transitions_by_level_normalized[0]])], []
+
+    def value(self):
+        arr, transitions = self.args
+        arrval = [argval(a) for a in arr]
+        curr_node = self.root_node
+        for v in arrval:
+            if (curr_node, v) in self.mapping:
+                curr_node = self.mapping[curr_node, v]
+            else:
+                return False
+        return curr_node == self.sink_node
+
+class Regular(GlobalConstraint):
+    """The values of the variables in 'array' correspond to a path in the automaton formed by the transition in 'transitions'.
+    The path starts in 'start' and ends in one of the ending states ('ends')
+    """
+    def __init__(self, array, transitions, start, ends):
+        array = flatlist(array)
+        if not all(isinstance(x, Expression) for x in array):
+            raise TypeError("the first argument of a regular constraint should only contain variables/expressions")
+        super().__init__("regular", [array, transitions, start, ends])
+        self.mapping = {}
+        for s, v, e in transitions:
+            self.mapping[(s, v)] = e
+
+    def decompose(self):
```

Comment:
some high-level docstring to get an idea / overview of how this decomposition works, so that the reader has a vague idea of its workings before delving into the code